scripts/create_deprecated_module.lean

```diff
@@ -0,0 +1,175 @@
+    CommandElabM Format := do
+  let msgStx := if customMessage.isEmpty then none else some <| Syntax.mkStrLit customMessage
+  let dateStx := Syntax.mkStrLit s!"{← Std.Time.PlainDate.now}"
+  let stx ← `(command|deprecated_module $[$msgStx]? (since := $dateStx))
+  liftCoreM <| PrettyPrinter.ppCategory `command stx
+
+/--
+The command `#create_deprecated_modules (n)? (comment)? (write)?` generates deprecated modules.
+
+Writing
+```lean
+#create_deprecated_modules 5 "These files are no longer relevant"
+```
+looks at the `lean` files in `Mathlib` that existed `4` commits ago
+(i.e. the commit that you see with `git log -5`) and that are no longer present.
+It shows how the corresponding deprecated modules should look like, using
+`"These files are no longer relevant"` as the (optional) comment.
+
+If the number of commits is not explicitly used, `#create_deprecated_modules` defaults to `2`,
+namely, the commit just prior to the current one.
+
+If the message is not explicitly used, `#create_deprecated_modules` defaults to
+`"Auto-generated deprecation"`.
+If you wish there to be no comment, use `#create_deprecated_modules 5 ""`.
+
+Note that the command applies the *same* comment to all the files that it generates.
+
+Finally, if everything looks correct, adding a final `write` actually generates the files:
+```lean
+#create_deprecated_modules 5 "These files are no longer relevant" write
+```
+-/
+syntax "#create_deprecated_modules" (ppSpace num)? (ppSpace str)? (&" write")? : command
+
+elab_rules : command
+| `(#create_deprecated_modules%$tk $[$nc:num]? $[$comment:str]? $[write%$write?]?) => do
+  let n := nc.getD (Syntax.mkNumLit "2") |>.getNat
+  let mut msgs := #[]
+  let getHash (n : Nat) := do
+    let log ← IO.Process.run {cmd := "git", args := #["log", "--pretty=oneline", s!"-{n}"]}
+    let some last := log.trim.splitOn "\n" |>.getLast? | throwError "Found no commits!"
+    let commitHash := last.takeWhile (!·.isWhitespace)
+    return commitHash
+  let getFilesAtHash (hash : String) := do
+    let files ← IO.Process.run
+      {cmd := "git", args := #["ls-tree", "-r", "--name-only", hash, "Mathlib/"]}
+    let h : Std.HashSet String := .ofList <| files.splitOn "\n"
+    return h
+  let currentHash ← getHash 1
+  let currentFiles ← getFilesAtHash currentHash
+  msgs := msgs.push m!"{currentFiles.size} files at the current hash {currentHash}\n"
+  let pastHash ← getHash n
+  let pastFiles ← getFilesAtHash pastHash
+  msgs := msgs.push m!"{pastFiles.size} files at the past hash {pastHash}\n"
+  let onlyPastFiles := pastFiles.filter fun fil ↦
+    fil.takeRight ".lean".length == ".lean" &&
+    !currentFiles.contains fil
+  let noFiles := onlyPastFiles.size
+  msgs := msgs.push
+    m!"{noFiles} Lean file{if noFiles == 1 then "" else "s"} in 'Mathlib' that no longer exist."
+  let deprecation← if let some cmt := comment then mkDeprecation cmt.getString else mkDeprecation
+  msgs := msgs.push ""
+  for fname in onlyPastFiles do
```

Comment:
-- Generate a module deprecation for each file that only existed in the past commit.

Reply:
Comment added.